### ISSUE TITLE
Chemistry and semantics/AT

### ISSUE THREAD

**Issue by NSoiffer** [w3c/mathml#93]
The notations used in chemical formulas are very similar to those used in math. People often use math editors to enter those formulas, which means they will show up as MathML on the web. The meaning/way a chemical formula is read is different than that of the corresponding math notation. E.g, ![image](https://user-images.githubusercontent.com/1545836/58515987-5fc0f900-815b-11e9-89a6-510479814005.png) might be read as "cap h sub 2 cap o" if it were math but if interpreted as chemistry it would be read as "h 2 o" or maybe even as "water". Similarly, → might be spoken as "right arrow" or "maps to" or some other math terminology in a mathematical context, but as "yields" in a chemistry context.

Chemical formulas often involve math notations and computations, along with units. So there is some mixing of math and chemistry. MathPlayer is able to read chemistry formulas. It has heuristics to try and intuit if the MathML is math or chemistry. If it determines it is chemistry, chemistry-specific rules are applied before falling back to more general math rules.

It seems like there are two potential places to indicate chemistry:
1. On the `math` element, perhaps with something like `role="chemistry"` depending on how semantics and presentation MathML is resolved (#64, #68, #79, #80). This would tell applications that the contents should be interpreted as a chemical formula unless overridden in by some child.
2. On the leaf elements such as the elements and various arrows that have specific meanings in chemistry. If a base of a script is tagged as being chemistry, then that would be lifted to the script elements so that ions, atomic numbers, etc., so that ions, etc., are read appropriately.
3. @TzviyaSiegman brought up an issue with how chemical compounds are spoken, specifically compounds that are rare and likely whose name is made up by the author. I'm not clear on how this relates to chemical formula semantics other than perhaps a requirement that there be a way to tag the compound with a name to be read (e.g, "sulfuric acid" or "water") when an alternative to reading out the details is desired. Hopefully she can clarify this...
4. @CarySupalo brought up an issue that "K" can be upper or lower case, italics or regular and he can't tell the difference using his AT. I know that "K" can mean the element Potassium, an equilibrium constant, or a temperature in Kelvin. I don't know what the other "k"s are. Hopefully he can clarify.

Note: There is a [braille chemistry code](http://www.brl.org/chemistry/rule07.html), but it appears that it uses Nemeth or UEB code for chemical formulas. _Someone more knowledgeable than me should verify this_. I don't know about non-English braille codes and chemistry. If chemical forumlas are brailled differently, then that is a strong argument for marking the `math` element with something to indicate this is chemistry.

**Comment by mhchem** [comment]
The German-speaking Braille system for chemistry ([source](https://www.chids.de/dachs/wiss_hausarbeiten/ChemieFuerBlinde_Ruhrmann.pdf#page=63)) is different from the BLR version mentioned above. It is a subset of German [mathematics braille](http://www.bskdl.org/download/mathematik/Mathematik%20Zeichenliste%20EB.pdf), so no special tagging would be needed for a MathML-to-Braille transformation. (Interestingly, this system loses the upright/italic distinction. But for chemistry, this is a problem only in very rare edge cases.)

**Comment by NSoiffer** [comment]
In https://github.com/mathml-refresh/mathml/issues/92#issuecomment-496987469, @mhchem wrote:

> From my experience, I would avoid semantic markup, i.e. giving each part of η²-C₂H₄ a description of why it is typeset as it is. The same notation (upright greek, dash, dot, ...) can have several very special chemical meanings, depending on the field of chemistry, with new meanings being added (and forgotten) all the time.

In the case of the Greek letter, would someone reading that normally just say "eta" or would they use some special name like "Frank's fudge factor" when reading it. If the latter, then a publisher may want to take the time to mark it up so that (for example) η is read as "Frank's fudge factor". That would mean MathML needs to have a way to allow that.

The same is question can be asked for the entire notation η²-C₂H₄. Would the author of that notation potentially have a special name for the entire expression? Would the "η", "η²", or "η²-" part be read specially? Or would a normal reader simply read that as something like "eta squared dash C 2 H 4" and know the special meanings from context?

**Comment by Steve-Noble** [comment]
Did we resolve the issue of how to disambiguate where MathML has been used to markup chemical formulas as opposed to mathematics? Some type of "subject" tag that would signal when chemistry is being presented could then be potentially intercepted by assistive technology applications and used to switch speech rules to chemistry, rather than relying on advanced heuristics being added by the AT developers, which will always be subject to some degree of error.

**Comment by dginev** [comment]
A subject tag is not a viable solution without an entire subject ontology that underpins what is and isn't covered by it.

Worse, in advanced materials such as research articles, subjects are interleaved - one can see chemical equations in some astrophysics articles, as one example.

We should prototype some chemical equations with the "intent" mechanism as designed thus far and see where the painpoints are, if any.

To that end, I have procured about twenty advanced chemical equation labels from the mineral wing of the Museum of Natural History in Vienna (was passing through this week) , and will be preparing some MathML snippets for the most challenging of them in the coming days.

**Comment by Steve-Noble** [comment]
Yes...I agree that some type of ontology would need to be defined. It is not entirely dissimilar to the problem with the HTML <lang> attribute when a web page author is including quotes of authors in various languages in the same web page. If the "intent" framework proposes to take care of that for differentiating between chemistry and other species of MathML expressions, then that would be great. 

I am sure there are plenty of times when web page authors want to include something like (for example) nuclear decay chemical formulas in one place on a page, and in another place showing something like an overall heat transfer coefficient equation to discuss nuclear power generation. I am sure there are plenty of examples where similar needs come up in engineering, chemistry and physics disciplines. It would seem to be a very common use case.

**Comment by polx** [comment]
But then... we should just use different "notations", i.e. different MathML.
Saying "this is chemistry here" will fail everytime you have stochiometry coefficients which are mathematical objects (e.g. random variables).

Or?
Also, what about the fields which live by mixing the fields? (bio-physics for example)

**Comment by NSoiffer** [comment]
The reason for having a `subject` attr is to allow for simpler remediation of existing content. It should _always_ be the case that one can use `intent` for chemistry notation or any other notation and it would override whatever defaults are in place at the time. For chemistry, having to use `intent` is a lot of overriding though. Probably beyond what a person could reliably do by hand for anything other than the simplest expressions.

@davidfarmer: have you looked into chemistry in PreTeXt? Do you just rely on the mchem (or some other) package?

I also should note that the [list of intent values](https://docs.google.com/spreadsheets/d/1EsWou1K5nxBdLPvQapdoA9h-s8lg_qjn8fJH64g9izQ/edit#gid=1358098730) has a "subject" column (probably at my urging), so that is a start on an ontology. I think it is overly specific for the needs of a `subject` attr. But that's also another way of saying I agree with @dginev that this needs more thought and more motivating examples.

**Comment by dginev** [comment]
> The reason for having a subject attr is to allow for simpler remediation of existing content.

I like the idea of making remediation as easy as possible, ideally just adding a single magic keyword for standard cases. The hard question is how to execute this cleanly, essentially to a point where `subject="X"` has a clear-cut standard interpretation of:

"use a well-defined resource of name X, which maps a collection of presentation MathML notations to intent expressions"

My concern is that we will need to develop these subject-specific catalogs ourselves and will fully enter the territory of what I call "the Content Dictionary problem". I won't expand on that here, but briefly - it would not be viable without a very serious curation+maintenance investment.

Otherwise, each AT will be at liberty to interpret these keywords in whichever way they consider best for the communities ("clients") they serve, leading to a non-standard outcome. Some AT will have a "Nitrogen" for "N", others won't. Worse, some AT may support "chemistry" while others may completely omit it. Thus no available guarantees for any group of stakeholders, be they authors, remediators or readers, unless they all work with (a specific version of) a toolchain of a single vendor. Which brings me back to suggesting to simply use `data-subject` until we know enough to standardize a solution that can be leveraged by all tools, and relied on by all stakeholders.

Even after such hypothetical ontology bootstrapping takes place, we would then need to provide a proper name resolution algorithm, which avoids ambiguity, or at least signals it to remediators. For example, adding a `subject="chemistry,physics"` to an expression that leverages multiple subject areas will immediately lead to conflicts between chemical elements and units. Example: `N` could be a "Newton" unit of force, or the chemical element Nitrogen. `K` can be Kelvin or Potassium, `C` can be Celsius or Carbon...

I'm currently more skeptical than optimistic about including such a mechanism at this stage.

> I also should note that the [list of intent values](https://docs.google.com/spreadsheets/d/1EsWou1K5nxBdLPvQapdoA9h-s8lg_qjn8fJH64g9izQ/edit#gid=1358098730) has a "subject" column (probably at my urging)

They were added at my sole discretion and mostly served for internal bookkeeping while compiling the list. At the time, I wanted us to get a feeling of how the notations are distributed in different areas, and how difficult the overloading problem is. I am unsure if they are useful beyond that - if we wanted a naming scheme, we ought to look toward somehow reusing large classification efforts such as the [MSC](https://en.wikipedia.org/wiki/Mathematics_Subject_Classification), [CCS](https://en.wikipedia.org/wiki/ACM_Computing_Classification_System), [PACS](https://en.wikipedia.org/wiki/Physics_and_Astronomy_Classification_Scheme), [SI](https://en.wikipedia.org/wiki/International_System_of_Units), ... but it gets overwhelming pretty quickly. The relative peace of K-12 taxonomy disappears rather abruptly once we get into higher ed materials.

**Comment by davidcarlisle** [comment]
I would say we have no possibility of defining ontologies covering entire subject areas in the available time (or ever) so making that a prerequisite is essentially the same as saying not to add the feature.

I am not convinced that the feature is needed, but if we do add it I think it should be something _much_ more lightweight. Essentially just some keywords possibly useful for a system to customise its defaults which are anyway going to be system specific. A system like @NSoiffer's mathcat even if it follows the (eventually written) `intent` spec, includes lots of heuristics and pattern matching to produce the best reading it can. These readings are never going to be the same across all systems. If a subject attribute gives it a hint to switch its defaults so an H with no intent ends up being inferred as being about Hydrogen rather than Homology, that is not clearly a bad thing and does not require an ontology detailing all possible differences between chemistry and mathematics.

A document could put intent on everything and avoid system specific defaulting. but it remains to be seen if that will happen in practice.

So I agree we shouldn't add any new attribute without more use cases and experiments, but I don't agree that defining a formal ontology would be a prerequisite of adding one.

**Comment by brucemiller** [comment]
I believe that our aim should be to make `intent` sufficiently expressive (along with an appropriate & extensible scheme for dictionaries) so that chemical formula, as well as math, can be annotated with intent.

We should probably think of (a future) `subject` attribute as being as an aid for remediation, as @NSoiffer said. More specifically, we should think of a subject as corresponding to a set of rules which match MathML subtrees and describe how to apply (standard) `intent` annotations to them.  This approach makes the feature a separate layer,  allow this it to be deferred, and experimented with (possibly even a purely server-side operation).

**Comment by dginev** [comment]
For @Steve-Noble and everyone else tracking here,

Following up on [my comment from June 17](https://github.com/w3c/mathml/issues/93#issuecomment-1158844568), here is my write-up of some advanced syntax used in chemical equations for mineralogy. 

I stumbled on them somewhat by accident in my travels this summer:

https://hackmd.io/@dginev/B1R3Xqup9

I don't understand all examples fully, but I am quite happy I can now offer this extended set of syntactic uses to the group.

Edit: here is a single example preview to motivate reading the full post.
<img width="700" src="https://user-images.githubusercontent.com/348975/182857397-ffc04846-352e-4210-b759-6bbd9d46d9a8.png">

**Comment by NSoiffer** [comment]
Very interesting notations!

In some email exchanges with various people on the Chem CG, some notations I found on the web (e.g, state of matter in subscript) were considered older notations and not current recommended practice. I have seen the use of variables in chemistry notations (e.g., a hydrocarbon chain).

I'll point the Chem CG to this issue and perhaps some of them will have some comments.

**Comment by NSoiffer** [comment]
I stumbled upon [another web page](https://www.nuclear-power.com/radioactive-decay/) with interesting notation. It is sort of a meta notation since it doesn't contain an actual chemical element. Clicking on [the image](https://nuclear-power.com/wp-content/uploads/2014/10/notation_nuclear_reactions.jpg) makes it a bit more readable.

One sample from the image:
![image](https://user-images.githubusercontent.com/1545836/183808185-03dd1f45-88a9-41b4-b5d2-41f113330093.png)

**Comment by mhchem** [comment]
That is not a new notation at all. That is using variables instead of names. The same thing as in 
![image](https://user-images.githubusercontent.com/16336692/183830045-7af0100d-c8f3-4366-bbb6-5fbdc9a484e6.png)
Variables should be in italic font. That is the case.
It also includes "variable elements" named X and X' (which do not stand for a number but for a chemical element). I am not aware of a clear rule, but mostly, those are written upright, as this eases interpretation.

**Comment by NSoiffer** [comment]
Probably resolved via `isa` (#426)
